### PR TITLE
_source search parameter needs to support modifiers

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
Added support for :above, :below, :contains and :missing _source search parameter modifiers.

### CODE REVIEW

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+import org.hl7.fhir.instance.model.api.IIdType;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIf;
+
+/**
+ * Test cases for _source search parameter.
+ */
+public abstract class SourceSearchParameterTestCases implements ITestDataBuilder.WithSupport {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java:34]: We use `Base` for abstract parent classes:
```suggestion
public abstract class BaseSourceSearchParameterTestCases implements ITestDataBuilder.WithSupport {
```

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+				"search source URI for Patient finds", "Patient?_source=http://host/0", pt0id);
+		myTestDaoSearch.assertSearchNotFound(
+				"search source URI for Patient - Observation not found", "Patient?_source=http://host/0", ob0id);
+	}
+
+	@Test
+	public void testSearchSource_withOrJoinedParameter_returnsUnionResultBundle() {
+		myTestDataBuilder.getSystemRequestDetails().setRequestId("a_request_id");
+
+		IIdType pt0id = createPatient(withSource(getFhirContext(), "http://host/0"), withActiveTrue());
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java:123]: `withSource()` should be able to use the FhirContext already bound to the builder (see `ITestDataBuilder.getFhirContext()`).  Please add a call without the first param, and use that.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java (hunk 1 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java]: Renamed

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java (hunk 2 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/SourceSearchParameterTestCases.java]: Thanks for suggestion. Applied changes

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java [hunk 3]
```diff
@@ -602,6 +603,43 @@ public static List<NameValuePair> translateMatchUrl(String theMatchUrl) {
 		return parameters;
 	}
 
+	/**
+	 * Creates list of sub URIs candidates for search with :above modifier
```
Comment (michaelabuckley) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java:607]: Nice docs.  Examples are always good.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java [hunk 4]
```diff
@@ -602,6 +603,43 @@ public static List<NameValuePair> translateMatchUrl(String theMatchUrl) {
...
+	 * Example input: http://[host]/[pathPart1]/[pathPart2]
+	 * Example output: http://[host], http://[host]/[pathPart1], http://[host]/[pathPart1]/[pathPart2]
+	 *
+	 * @param theUri String URI parameter
+	 * @return List of URI candidates
+	 */
+	public static List<String> getAboveUriCandidates(String theUri) {
```
Comment (michaelabuckley) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java:614]: This code is complicated enough for a unit test.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java [hunk 5]
```diff
@@ -602,6 +603,43 @@ public static List<NameValuePair> translateMatchUrl(String theMatchUrl) {
...
+
+			if (uri.getScheme() == null || uri.getHost() == null) {
+				throwInvalidRequestExceptionForNotValidUri(theUri, null);
+			}
+			StringBuilder sb =
+					new StringBuilder().append(uri.getScheme()).append("://").append(uri.getHost());
+
+			candidates.add(sb.toString());
+
+			String[] pathParts = uri.getPath().split("/");
```
Comment (michaelabuckley) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java:627]: You might try using java.nio.file.Path.parent() to do this without all this string work.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/SourcePredicateBuilder.java [hunk 6]
```diff
@@ -53,6 +69,46 @@ public Condition createPredicateSourceUri(String theSourceUri) {
...
+						FunctionCall upperFunction = new FunctionCall("UPPER");
+						upperFunction.addCustomParams(myColumnSourceUri);
+						String normalizedString = StringUtil.normalizeStringForSearchIndexing(theSourceUri);
+						String containsLikeExpression = createLeftAndRightMatchLikeExpression(normalizedString);
+						return BinaryCondition.like(upperFunction, generatePlaceholder(containsLikeExpression));
+					} else {
+						throw new MethodNotAllowedException(
+								Msg.code(2417) + ":contains modifier is disabled on this server");
+					}
+				default:
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/SourcePredicateBuilder.java:101]: trivial: each case is big enough to extract into a small function.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java [hunk 7]
```diff
@@ -270,6 +174,41 @@ public void deleteWithSource() {
...
+			assertEquals(Msg.code(2417) + ":contains modifier is disabled on this server", e.getMessage());
+		}
+	}
+
+	@ParameterizedTest
+	@CsvSource({
+		"invalid_uri",
+		"http://some-source/ with_invalid_uri"
+	})
+	public void testSearchSource_withAboveModifierAndInvalidURI_throwsException(String theUri) {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java:195]: trivial: For single strings, @ValueSource is safer.

Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java:195]: Nice to see these cleaned away.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/dao/TestDaoSearch.java [hunk 8]
```diff
@@ -72,9 +72,6 @@ TestDaoSearch testDaoSearch(
 		}
 	}
 
-	@Autowired
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/dao/TestDaoSearch.java:75]: problem: not always present.  Add "optional=true"

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java [hunk 9]
```diff
@@ -404,6 +409,8 @@ default ICreationArgument[] asArray(ICreationArgument theIBaseResourceConsumer)
 	}
 
 	interface Support {
+		SystemRequestDetails getSystemRequestDetails();
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java:412]: minor: we have been trying to keep this interface generic so we could bind it to GenericClient.  But RequestDetails is an internal Dao class.  I would rather have an explicit `setRequestId` that we could support in both places.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java [hunk 10]
```diff
@@ -191,6 +192,10 @@ default ICreationArgument withSource(FhirContext theContext, String theSource) {
 		return t -> MetaUtil.setSource(theContext, ((IBaseResource)t).getMeta(), theSource);
 	}
 
+	default ICreationArgument withSource(String theSource) {
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java:195]: thank you.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/BaseSourceSearchParameterTestCases.java [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+				p3Id);
+
+		myTestDaoSearch.assertSearchFinds(
+				"search matches one sources (intersection with family SearchParameter)",
+				"Patient?_source:contains=Another-Source,some-source&family=Family,YourFamily",
+				p3Id);
+
+		myTestDaoSearch.assertSearchNotFound(
+				"search returns empty bundle (contains with missing=true)",
+				"Patient?_source:contains=Another-Source,some-source&_source:missing=true",
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/BaseSourceSearchParameterTestCases.java:183]: question:  Does this test cover the complicated bit in SearchBuilder with two different join types (LEFT vs INNER)?  That was the most complciated bit you added.

**Review by michaelabuckley — APPROVED** [review]
Looks good.  Some minor suggestions and questions.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java (hunk 4 above)
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java:614]: Agree here. Added tests for `getAboveUriCandidates` to UrlUtilTest.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java (hunk 5 above)
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java]: Thank You for suggestion. With `java.nio.file.Path` usage code is more readable.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/dao/TestDaoSearch.java (hunk 8 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/dao/TestDaoSearch.java:75]: Thanks. Added

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java (hunk 7 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java]: test moved to UrlUtilTest with `@ValueSource` usage

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java (hunk 3 above)
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/UrlUtil.java:607]: Thank You!

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/BaseSourceSearchParameterTestCases.java (hunk 11 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/BaseSourceSearchParameterTestCases.java:183]: Yes, this test along with `testSearchSource_withMissingModifierTrue_returnsEmptySources` covers two different join types (LEFT vs INNER) cases. To check if different types of joins are working correctly we need to execute search with `_source:missing=true` with intersection on another SearchParameter (I also added one more test case to `testSearchSource_withMissingModifierTrue_returnsEmptySources` to check that)

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java (hunk 7 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SourceTest.java]: test moved to UrlUtilTest with @ValueSource usage

**Pull request merged**